File: ContainerWithMostWater.py
"""
11. Container With Most Water
Given n non-negative integers a1, a2, ..., an, where each represents a point at coordinate (i, ai). n vertical lines are drawn such that the two endpoints of line i is at (i, ai) and (i, 0). Find two lines, which together with x-axis forms a container, such that the container contains the most water.
Note: You may not slant the container.
"""
# O(n)
# O(1)
class Solution(object):
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        res = 0
        i, j = 0, len(height) - 1
        while i < j:
            res = max(res, (j - i) * min(height[i], height[j]))
            if height[i] <= height[j]:
                i += 1
            else:
                j -= 1
        return res

if __name__ == "__main__":
    height = [1,3,2,5,1,7]
    print(Solution().maxArea(height))

File: Java/Container With Most Water.java

public class maxArea {

    public int maxArea(int[] height) {
        int i = 0;
        int j = height.length - 1;
        int res = 0;
        while(i < j) {
            res = Math.max(res, (j - i) * Math.min(height[i], height[j]));
            if(height[i] < height[j]) {i++;}
            else {j--;}
        }
        return res;
    }

}
